tests/unit/test_shared_release_info.py

```diff
@@ -0,0 +1,67 @@
+import pytest
+
+from src.shared.release_info import *
+
+
+def test_get_revision_to_release_plain():
+    all_releases = {
+        "3.8-20.04": {
+            "end-of-life": "2025-03-31T00:00:00Z",
+            "edge": {"target": "43"},
+            "stable": {"target": "43"},
+            "candidate": {"target": "43"},
+            "beta": {"target": "43"},
+        }
+    }
+    assert get_revision_to_released_tags(all_releases) == {
+        43: [
+            "3.8-20.04_beta",
+            "3.8-20.04_candidate",
+            "3.8-20.04_edge",
+            "3.8-20.04_stable",
+        ]
+    }
+
+
+def test_get_revision_to_release_circular():
+    all_releases = {
+        "1.19.0-22.04": {
+            "end-of-life": "2024-11-26T00:00:00Z",
+            "stable": {"target": "1"},
+            "candidate": {"target": "1.19.0-22.04_beta"},
+            "beta": {"target": "1.19.0-22.04_candidate"},
+            "edge": {"target": "5"},
+        }
+    }
+
+    with pytest.raises(BadChannel) as excinfo:
+        get_revision_to_released_tags(all_releases)
+    assert (
+        "Tag 1.19.0-22.04_candidate was caught in a circular dependency, following tags that follow themselves. Cannot pin a revision."
```

Comment:
nit:   I think we over constraining the tests to check for this string. Maybe it is good enough that we assert BadChannel is raised. We could also then create a new exception class for circular dependencies if we need to single out this exact scenario.